lang/cs/Org.Apache.REEF.Common/metrics/MetricsSystem/MetricsSystem.cs

```diff
@@ -0,0 +1,307 @@
+                }
+                PublishMetrics(all);
+            }
+        }
+
+        /// <summary>
+        /// Shuts down the metrics system. Call Stop() befor ShutDown()
+        /// </summary>
+        /// <returns>True if shutdown is successful.</returns>
+        public bool ShutDown()
+        {
+            lock (_lock)
+            {
+                _shutDown = true;
+                foreach (var sink in _sinks)
+                {
+                    sink.Value.ShutDown();
+                }
+                _sinks.Clear();
+                _sources.Clear();
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Function called to take snapshot from sources and put the in the sink queue. 
+        /// This method is either called periodically or by user via. <see cref="PublishMetricsNow"/>.
+        /// </summary>
+        private void PublishMetrics(bool all)
+        {
+            lock (_lock)
+            {
+                if (_sinks.Count != 0 && _sources.Count != 0)
+                {
+                    _sourceSnapshots.Clear();
+                    foreach (var source in _sources)
+                    {
+                        if (_sourceFilter.AcceptsName(source.Key))
+                        {
+                            _metricsCollector.Clear();
+                            _sourceSnapshots.Add(source.Value.GetMetrics(_metricsCollector, all));
+                        }
+                    }
+                    foreach (var sink in _sinks)
+                    {
+                        sink.Value.PutMetricsInQueue(_sourceSnapshots);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposable class for IObservable interface. Allows sinks to unregister themselves from 
+        /// the metrics system.
+        /// </summary>
+        private class SinkDisposableHandler : IDisposable
+        {
+            private readonly Dictionary<int, SinkHandler> _sinkHandlers;
+            private readonly int _key;
+            private readonly object _lock;
+
+            public SinkDisposableHandler(Dictionary<int, SinkHandler> sinkHandlers, int key, object lockObj)
+            {
+                _sinkHandlers = sinkHandlers;
+                _key = key;
+                _lock = lockObj;
+            }
+
+            public void Dispose()
+            {
+                lock (_lock)
+                {
+                    if (_sinkHandlers.ContainsKey(_key))
+                    {
+                        if (!_sinkHandlers[_key].IsStopped && !_sinkHandlers[_key].IsShutDown)
+                        {
+                            _sinkHandlers[_key].Stop();
+                        }
+                        _sinkHandlers.Remove(_key);
+                    }
+                    else
+                    {
+                        Logger.Log(Level.Warning, "Sink to remove from metrics system is not present");
```

Comment:
We should probably log identifier of sink not present, otherwise this log is not helpful for debug